activerecord/lib/active_record/insert_all.rb

```diff
@@ -0,0 +1,185 @@
+# frozen_string_literal: true
+
+module ActiveRecord
+  class InsertAll
+    def initialize(model, inserts, **options)
+      raise ArgumentError, "inserts should be an array" unless inserts.is_a?(Array)
+      @model, @connection, @options = model, model.connection, options
+      @keys, @values_list = connection.prepare_keys_and_values_for_insert(inserts, model.table_name)
+    end
+
+    def execute
+      if values_list.empty?
+        ActiveRecord::Result.new([], [])
+      else
+        connection.exec_query to_sql, "Bulk Insert"
+      end
+    end
+
+    def to_sql
+      sql_builder.build_sql(
+        model,
+        keys,
+        values_list,
+        returning: returning,
+        on_duplicate: on_duplicate,
+        unique_by: unique_by,
```

Comment:
Thinking we're missing an explicit InsertAllStrategy object here that can hold these rules. Then you can pull in convenience getters like skip_duplicates? as well.